swiftsimio/visualisation.py

```diff
@@ -0,0 +1,195 @@
+    """
+    inverse_H = 1.0 / (h * kernel_gamma)
+    ratio = r * inverse_H
+
+    kernel = 0.0
+
+    if ratio < 1.0:
+        ratio_2 = ratio * ratio
+        ratio_3 = ratio_2 * ratio
+
+        if ratio < 0.5:
+            kernel += 3.0 * ratio_3 - 3.0 * ratio_2 + 0.5
+
+        else:
+            kernel += -1.0 * ratio_3 + 3.0 * ratio_2 - 3.0 * ratio + 1.0
+
+        kernel *= kernel_constant * inverse_H * inverse_H
+
+    return kernel
+
+
+@jit(nopython=True)
+def pair_max(a: int32, b: int32) -> int32:
+    if a > b:
+        return a
+    else:
+        return b
+
+
+@jit(nopython=True)
+def pair_min(a: int32, b: int32) -> int32:
+    if a < b:
+        return a
+    else:
+        return b
+
+
+@jit(nopython=True, fastmath=True)
+def scatter(x: float32, y: float32, m: float32, h: float32, res: int) -> ndarray:
+    """
+    Creates a scatter plot of:
+
+    + x: the x-positions of the particles. Must be bounded by [0, 1].
+    + y: the y-positions of the particles. Must be bounded by [0, 1].
+    + m: the masses (or otherwise weights) of the particles
+    + h: the smoothing lengths of the particles
+    + res: the number of pixels.
+    
+    This ignores boundary effects.
+
+    Note that explicitly defining the types in this function allows
+    for a 25-50% performance improvement. In our testing, using numpy
+    floats and integers is also an improvement over using the numba ones.
+    """
+    # Output array for our image
+    image = zeros((res, res), dtype=float32)
+    maximal_array_index = int32(res)
+
+    # Change that integer to a float, we know that our x, y are bounded
+    # by [0, 1].
+    float_res = float32(res)
+    pixel_width = 1.0 / float_res
+
+    # If the kernel width is smaller than this, we drop to just PIC method
+    drop_to_single_cell = pixel_width * 0.5
+
+    # Pre-calculate this constant for use with the above
+    inverse_cell_area = res * res
+
+    for x_pos, y_pos, mass, hsml in zip(x, y, m, h):
+        # Calculate the cell that this particle
+        particle_cell_x = int32(float_res * x_pos)
+        particle_cell_y = int32(float_res * y_pos)
+
+        # SWIFT stores hsml as the FWHM.
+        kernel_width = kernel_gamma * hsml
+
+        if kernel_width < drop_to_single_cell:
+            # Easygame, gg
+            image[particle_cell_x, particle_cell_y] += mass * inverse_cell_area
+        else:
+            # The number of cells that this kernel spans
+            cells_spanned = int32(1.0 + kernel_width * float_res)
+
+            # Now we loop over the square of cells that the kernel lives in
+            for cell_x in range(
+                # Ensure that the lowest x value is 0, otherwise we'll segfault
+                pair_max(0, particle_cell_x - cells_spanned),
+                # Ensure that the highest x value lies within the array bounds,
+                # otherwise we'll segfault (oops).
+                min(particle_cell_x + cells_spanned, maximal_array_index),
```

Comment:
why not use pair min since you defined it?

Reply:
That's an excellent point. I have no idea how this worked before.